src/main/java/org/apache/ibatis/builder/annotation/AbstractProviderSqlSource.java

```diff
@@ -0,0 +1,99 @@
+/**
+ *    Copyright 2009-2017 the original author or authors.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+package org.apache.ibatis.builder.annotation;
+
+import org.apache.ibatis.builder.BuilderException;
+import org.apache.ibatis.builder.SqlSourceBuilder;
+import org.apache.ibatis.mapping.BoundSql;
+import org.apache.ibatis.mapping.SqlSource;
+import org.apache.ibatis.reflection.ParamNameResolver;
+import org.apache.ibatis.session.Configuration;
+
+import java.lang.reflect.Method;
+
+/**
+ * @author Clinton Begin
+ * @author Kazuki Shimizu
+ * @author liuzh
+ * @since 3.4.6
+ */
+public abstract class AbstractProviderSqlSource implements SqlSource {
+
+  protected final Configuration configuration;
+  protected final SqlSourceBuilder sqlSourceParser;
+  protected final Class<?> providerType;
+  protected Method providerMethod;
+  protected String[] providerMethodArgumentNames;
+  protected Class<?>[] providerMethodParameterTypes;
+  protected ProviderContext providerContext;
+  protected Integer providerContextIndex;
+
+  /**
+   * @since 3.4.6
+   */
+  public AbstractProviderSqlSource(Configuration configuration, Object provider, Class<?> mapperType, Method mapperMethod) {
+    String providerMethodName;
+    try {
+      this.configuration = configuration;
+      this.sqlSourceParser = new SqlSourceBuilder(configuration);
+      this.providerType = (Class<?>) provider.getClass().getMethod("type").invoke(provider);
+      providerMethodName = (String) provider.getClass().getMethod("method").invoke(provider);
+
+      for (Method m : this.providerType.getMethods()) {
+        if (providerMethodName.equals(m.getName())) {
+          if (m.getReturnType() == String.class) {
+            if (providerMethod != null){
+              throw new BuilderException("Error creating SqlSource for SqlProvider. Method '"
```

Comment:
This nested `if` should be changed into a guard and the nesting removed:

```java
Arrays.stream(this.providerType.getMethods())
.filter(m -> providerMethodName.equals(m.getName()))
.filter(m -> m.getReturnType.equals(String.class))
.filter(ignored -> providerMethod != null)
.ifPresent(b -> throw new BuilderException(...)
```

I have seen java 8 code in the project so I know it's supported.

Reply:
Java 8 is not supported in the library - currently we are 1.6 compatible.  The Java 8 stuff is in the tests only.

But yeah - that is a terrible nested if statement.

Reply:
This abstract class is extracted from [ProviderSqlSource](https://github.com/mybatis/mybatis-3/blob/master/src/main/java/org/apache/ibatis/builder/annotation/ProviderSqlSource.java), and this part of the code is not modified.